include/nbl/builtin/hlsl/bxdf/base/cook_torrance_base.hlsl

```diff
@@ -172,79 +182,109 @@ struct SCookTorrance
+            {
+                return rr(NdotTorR, rcpOrientedEta);
+            }
+            bxdf::ReflectRefract<scalar_type> rr;
+            scalar_type NdotTorR;
+        };
         bxdf::ReflectRefract<scalar_type> rr;
         rr.refract = r;
-        ray_dir_info_type L = V.reflectRefract(rr, transmitted, rcpEta.value[0]);
+        reflect_refract_wrapper rrw;
+        rrw.rr = rr;
+        rrw.NdotTorR = LdotH;
+        ray_dir_info_type L = V.template reflectRefract<reflect_refract_wrapper>(rrw, transmitted, rcpEta.value[0]);
 
         const vector3_type T = interaction.getT();
         const vector3_type B = interaction.getB();
-        const vector3_type _N = interaction.getN();
 
         // fail if samples have invalid paths
-        const vector3_type Ldir = L.getDirection();
-        const scalar_type LdotH = hlsl::dot(Ldir, H);
-        if ((ComputeMicrofacetNormal<scalar_type>::isTransmissionPath(VdotH, LdotH) != transmitted) || (LdotH * hlsl::dot(_N, Ldir) < scalar_type(0.0)))
+        const scalar_type NdotL = scalar_type(2.0) * VdotH * localH.z - localV.z;
+        if ((ComputeMicrofacetNormal<scalar_type>::isTransmissionPath(NdotV, NdotL) != transmitted))
             L.makeInvalid(); // should check if sample direction is invalid
         else
-            cache = anisocache_type::create(VdotH, Ldir, H, T, B, _N, transmitted);
+        {
+            fresnel::OrientedEtas<monochrome_type> orientedEta = fresnel::OrientedEtas<monochrome_type>::create(scalar_type(1.0), hlsl::promote<monochrome_type>(fresnel.getRefractionOrientedEta()));
+            assert(ComputeMicrofacetNormal::isValidMicrofacet(transmitted,cache.getVdotL(),localH.z,orientedEta));
+            cache.fillTangents(T, B, H);
+        }
 
         return sample_type::create(L, T, B, _N);
     }
-    template<typename C=bool_constant<!IsAnisotropic>, typename D=bool_constant<!IsBSDF> >
-    enable_if_t<C::value && !IsAnisotropic && D::value && !IsBSDF, sample_type> generate(NBL_CONST_REF_ARG(isotropic_interaction_type) interaction, const vector2_type u, NBL_REF_ARG(isocache_type) cache)
-    {
-        anisocache_type aniso_cache;
-        sample_type s = generate(anisotropic_interaction_type::create(interaction), u, aniso_cache);
-        cache = aniso_cache.iso_cache;
-        return s;
-    }
-    template<typename C=bool_constant<!IsAnisotropic>, typename D=bool_constant<IsBSDF> >
-    enable_if_t<C::value && !IsAnisotropic && D::value && IsBSDF, sample_type> generate(NBL_CONST_REF_ARG(isotropic_interaction_type) interaction, const vector3_type u, NBL_REF_ARG(isocache_type) cache)
+    template<typename C=bool_constant<!IsAnisotropic>, typename T=conditional_t<IsBSDF, vector3_type, vector2_type> >
+    enable_if_t<C::value && !IsAnisotropic, sample_type> generate(NBL_CONST_REF_ARG(isotropic_interaction_type) interaction, const T u, NBL_REF_ARG(isocache_type) cache)
```

Comment:
you can use `conditional_t` directly in a function signature, there's no need to add another template parameter `T`

P.S. also try to use NBL_FUNC_REQUIRES instead of the `enable_if_t`